src/platform_impl/linux/wayland/seat/pointer/mod.rs

```diff
@@ -0,0 +1,321 @@
+    relative_pointer: Option<ZwpRelativePointerV1>,
+
+    /// Confined pointer.
+    confined_pointer: Rc<RefCell<Option<ZwpConfinedPointerV1>>>,
+}
+
+impl Pointers {
+    pub(super) fn new(
+        seat: &Attached<WlSeat>,
+        theme_manager: &ThemeManager,
+        relative_pointer_manager: &Option<Attached<ZwpRelativePointerManagerV1>>,
+        pointer_constraints: &Option<Attached<ZwpPointerConstraintsV1>>,
+        modifiers_state: Rc<RefCell<ModifiersState>>,
+    ) -> Self {
+        let confined_pointer = Rc::new(RefCell::new(None));
+        let pointer_data = Rc::new(RefCell::new(PointerData::new(
+            confined_pointer.clone(),
+            pointer_constraints.clone(),
+            modifiers_state,
+        )));
+        let pointer = theme_manager.theme_pointer_with_impl(
+            seat,
+            move |event, pointer, mut dispatch_data| {
+                let winit_state = dispatch_data.get::<WinitState>().unwrap();
+                handlers::handle_pointer(pointer, event, &pointer_data, winit_state);
+            },
+        );
+
+        // Setup relative_pointer if it's available.
+        let relative_pointer = match relative_pointer_manager.as_ref() {
+            Some(relative_pointer_manager) => {
+                Some(init_relative_pointer(&relative_pointer_manager, &*pointer))
+            }
+            None => None,
+        };
+
+        Self {
+            pointer,
+            relative_pointer,
+            confined_pointer,
+        }
+    }
+}
+
+impl Drop for Pointers {
+    fn drop(&mut self) {
+        // Drop relative pointer.
+        if let Some(relative_pointer) = self.relative_pointer.take() {
+            relative_pointer.destroy();
+        }
+
+        // Drop confined pointer.
+        if let Some(confined_pointer) = self.confined_pointer.borrow_mut().take() {
+            confined_pointer.destroy();
+        }
+
+        // Drop the pointer itself in case it's possible.
+        if self.pointer.as_ref().version() >= 3 {
+            self.pointer.release();
+        }
+    }
+}
+
+pub(super) fn init_relative_pointer(
+    relative_pointer_manager: &ZwpRelativePointerManagerV1,
+    pointer: &WlPointer,
+) -> ZwpRelativePointerV1 {
+    let relative_pointer = relative_pointer_manager.get_relative_pointer(&*pointer);
+    relative_pointer.quick_assign(move |_, event, mut dispatch_data| {
+        let winit_state = dispatch_data.get::<WinitState>().unwrap();
+        handlers::handle_relative_pointer(event, winit_state);
+    });
+
+    relative_pointer.detach()
+}
+
+pub(super) fn init_confined_pointer(
+    pointer_constraints: &Attached<ZwpPointerConstraintsV1>,
+    surface: &WlSurface,
+    pointer: &WlPointer,
+) -> ZwpConfinedPointerV1 {
+    let confined_pointer =
+        pointer_constraints.confine_pointer(surface, pointer, None, Lifetime::Persistent.to_raw());
+
+    confined_pointer.quick_assign(move |_, _, _| {
+        // This space intentionally left blank.
```

Comment:
That basically assign callback to an wayland object and plumb them to a global filter you're using. I'm not sure that it'll work without it, I can try though.

Reply:
If you don't do quick_assign the events will be delivered to fallback closure, which we use unreachable, since events are not belong to anything...

Reply:
By "we use unreachable" you mean the winit code or sctk? That seems like a bad idea, why wouldn't the default closure just be empty?

Unreachable **should be unreachable**. If there's any way that it could be reached, it is not unreachable.

Reply:
>By "we use unreachable" you mean the winit code or sctk? That seems like a bad idea, why wouldn't the default closure just be empty?

In winit's sync_roundtrips and dispatches, The point of unreachable in winit is that we **always** assign callbacks to Wayland objects, so this code is **indeed** unreachable, if you decide to not assign callbacks to an objects the code can be reached, but I consider it a bug in winit. The code is in winit and closures that are responsible for those are the ones we provide to sync roundtrip and wayland source.

Reply:
Why do we force ourselves to always add callbacks? Isn't that just unnecessary extra work?

Reply:
>Why do we force ourselves to always add callbacks? Isn't that just unnecessary extra work?

it's not, if you don't add it you'll execute other callback aka fallback callback I've linked to you. So you're not changing by that anything. In general every object has a callback associated with it, in case you won't provide one it'll go into some other callback like the one in a sync round trip.

Reply:
Do we actually define a callback for every single object, or do we rely on the fallback in some case? If it just goes back to a fallback callback, isn't that good enough?

Reply:
>Do we actually define a callback for every single object, or do we rely on the fallback in some case? If it just goes back to a fallback callback, isn't that good enough?

We define for everything, in general this object is the only one where we provided an empty callback, since we don't actually care about events it gets. The goal of using `unreachable` was to catch bugs if someone define something that ended up not having a callback and so one could catch it during review, since you'll in that handler with `unreachable!`. But I'm also fine providing empty callbacks for sync roundtrips.